Xeno End-Shift Conditions

TheSilverNuke opened the pull request:
Reworks xenos to function just a bit better. At the moment I'm putting it on the Git to get feedback. I've received positive vibes from players on the servers when asking about it but public review is always nice.

## Why It's Good For The Game

We've all had moments like these:

![2020-05-02 05_44_39-Window](https://user-images.githubusercontent.com/36066032/80864471-1ff27380-8c40-11ea-9f1a-d7ef060466b8.png)
![2019-12-09 21_52_45-_tg_Station Sybil  ENGLISH   US-WEST   100% LAG FREE _ Feline Frying Factory 14](https://user-images.githubusercontent.com/36066032/80864543-98593480-8c40-11ea-89a5-1b90e72218c2.png)

I figure it's better to at least let the xenos be able to grow a little bit, rather than having the shuttle be called at the slightest hint of an infestation. At least with this method more players will have a chance at encountering a xeno instead of never seeing them because one person called it out.

(Why would Centcom allow xenos on their station anyway if they'll stop even revs, gangsters, or a blob?)

## About The Pull Request

This will create a hostile environment while there is a Queen still kicking. Killing the queen will allow for the shuttle to leave. This should be fine since the severed matriarchy effect lasts for a bit more than two minutes, which is how long it takes for a shuttle to depart.

At the moment I'm struggling a bit as I'm a new coder. There's a few things that I'm not sure how to do so here's what's left.

- [x] Make the station actually explode. Delta does still trigger at the moment and ends the round, but the self destruct  never detonates. 
- [x] Add a GPS. I haven't done this yet.
- [x] Clean up code

## Changelog
:cl:
add: GPS signal to track the Queen
add: When a queen is alive and kicking on the station it prevents the shuttle from leaving
balance: If the xenos manage to completely overwhelm the station and no crew are left there's no need for admin intervention to end the round. A Delta alert will automatically trigger and self destruct the station.
balance: Minimum number of players for the alien infestation event to trigger has been increased
/:cl:

TheSilverNuke committed:
Xeno Overhaul

Reworks xenos to function just a bit better. WIP.

nemvar commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -10,9 +10,74 @@
 	layer = LARGE_MOB_LAYER //above most mobs, but below speechbubbles
 	pressure_resistance = 200 //Because big, stompy xenos should not be blown around like paper.
 	butcher_results = list(/obj/item/reagent_containers/food/snacks/meat/slab/xeno = 20, /obj/item/stack/sheet/animalhide/xeno = 3)
+	gps_name = "Hissing Signal"
```
To add a gps signal to the queen you just slap
```
AddComponent(/datum/component/gps, "Hissing Signal")
```
Into /mob/living/carbon/alien/humanoid/royal/queen/Initialize()

TheSilverNuke replied:
Nice! Thanks.

HomocidlePancakes commented:
Here’s a better solution:
https://github.com/yogstation13/Yogstation/pull/6350

In a later PR, xeno queen got the ability to remove shuttle lockdown by slashing at the shuttle console.

nemvar commented:
It's also important to consider WHY people are calling the shuttle if they see the slightest hint of a xeno and that's because it just isn't fun to play against them. I don't see how this fixes that problem.

TheSilverNuke commented:
> 
> 
> It's also important to consider WHY people are calling the shuttle if they see the slightest hint of a xeno and that's because it just isn't fun to play against them. I don't see how this fixes that problem.

I think a large part of the problem (aside from the neurotoxin spit, speed, and facehugger stun) is that they're like swarmers. They just keep coming back.

I believe by adding that GPS signal it should firstly allow for anyone who already was watching the GPS to be able to identify them immediately and prevent subsequent Queens from establishing too firm a hold on the station if one was already slain.

Qustinnus commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -10,9 +10,74 @@
+
+/mob/living/carbon/alien/humanoid/royal/queen/Life()
+	..()
+	if(client)
```
Early returns

trollbreeder replied:
Early returns are used for when your conditional is for the entire proc (checking for a client in this case). To do this, check for the opposite, **if there is no client**, and return if that is the case.
```
if(!client)
	return
```

Qustinnus replied:
they know what early returns are they use them in other moments

Qustinnus commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -10,9 +10,74 @@
+
+/mob/living/carbon/alien/humanoid/royal/queen/proc/nuke_it_from_orbit()
+	nuking = TRUE
+	sleep(50)
```
Don't use sleep. Use timers

TheSilverNuke replied:
In this context why is it better to use timers instead of sleep?

Qustinnus replied:
sleep is a byond inbuilt we don't use (except in very specific cases), we generally prefer using our own timer system.

Qustinnus commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -10,9 +10,74 @@
+		life_ticks_to_wait = initial(life_ticks_to_wait)
+		var/living_humans = 0
+		var/total_humans = length(GLOB.human_list)
+		for(var/H in GLOB.human_list)
```
Why not loop the player list and check if theyrr human instead

TheSilverNuke replied:
How would I write that?

Qustinnus commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -10,9 +10,74 @@
+		if(living_humans < total_humans/10 && !nuking && !neutralized)
+			INVOKE_ASYNC(src, .proc/nuke_it_from_orbit)
+
+/mob/living/carbon/alien/humanoid/royal/queen/proc/nuke_it_from_orbit()
```
This proc has no business being on the queen herself

TheSilverNuke replied:
Where would I put that instead?

Qustinnus replied:
most preferably some kind of over-arching manager for the xeno infestation, which I'm guessing doesnt exist because its not really a gamemode.

Qustinnus commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -10,9 +10,74 @@
+		nuking = FALSE
+		neutralized = TRUE
+
+/proc/xenomorph_nuke(where)
```
This shouldnt be a global proc

Qustinnus commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -10,9 +10,74 @@
+	if(!client)
+		SSshuttle.clearHostileEnvironment(src) //the queen has officially gone braindead, so we can allow the shuttle to leave
+
+/mob/living/carbon/alien/humanoid/royal/queen/Life()
```
Also this entire check has no business being on the queen either.

TheSilverNuke replied:
Thanks a ton for the input. I do appreciate it. I've actually been tinkering off and on with this PR and getting a bit of help from coderbus for awhile. I'll go ahead and work on moving the things you mentioned.

TheSilverNuke committed:
Update queen.dm

Archanial commented:
Xenos should be their own gamemode instead of being random spawn. 
But if you want to keep them as spawn they should be balanced propertly. One hit stun and then being facehugged for 5 minutes until you die isn't fun. Xenos wipe crew too easy most of the time. That's why shuttle is called most of the time.

TheSilverNuke commented:
> 
> 
> Xenos should be their own gamemode instead of being random spawn.
> But if you want to keep them as spawn they should be balanced propertly. One hit stun and then being facehugged for 5 minutes until you die isn't fun. Xenos wipe crew too easy most of the time. That's why shuttle is called most of the time.

I do agree some things should be reworked, such as stun times. Various elements have been seeing stun reworks recently and xenos shouldn't be an exception. If we were to make it it's own gamemode though that would probably be a little too close to what CM and TGCM currently offers.

Archanial commented:
> 
> 
> > Xenos should be their own gamemode instead of being random spawn.
> > But if you want to keep them as spawn they should be balanced propertly. One hit stun and then being facehugged for 5 minutes until you die isn't fun. Xenos wipe crew too easy most of the time. That's why shuttle is called most of the time.
> 
> I do agree some things should be reworked such as stun times. Various elements have been seeing stun reworks recently and xenos shouldn't be an exception. If we were to make it it's own gamemode though that would probably be a little too close to what CM and TGCM currently offers.

CM and TGMC are whole different games and I don't see how xeno infestation on space station is close to marines storming xeno hideouts.

Tlaltecuhtli commented:
xenos are aids as they steamroll anyone with poor equipment, if you arent wearing riot armor and scatter shotgun you are insta fucked by thrown huggers or shove stun locked and killed, multiple xenos can easly take care of any amount of gaming gear or sec, if it wasnt for early calls/spotting 90% of xeno round would be a xeno victory and they dont even end the round if they managed to prevented shuttle call and even DEATHSQUAD cant deal with a xeno infestation 
leave xenos be as evil bogeyman  that happen rarely

FloranOtten commented:
There's a reason the shuttle is called. A good xeno queen is undefeatable.  They just keep coming back, they're fast and stealthy and unbeatable in melee. Even if you kill the queen with a lucky maxcap, any ol' drone can pop back in as a queen after 2 minutes (god help you if you killed them before the shuttle arrives)

bluetshirtguy commented:
can we not have another midround antag that forces the shuttle to stay? xenonormphs were meant to end rounds if out of control not stop them from ending until a winner is decided.

TheSilverNuke commented:
> 
> 
> can we not have another midround antag that forces the shuttle to stay? xenonormphs were meant to end rounds if out of control not stop them from ending until a winner is decided.

If people stop kneejerk calling the shuttle every time the x-word pops up, sure.

Shadowflame909 commented:
Reminder that these terrible nerds have a chance to spawn in xenobiology at roundstart. So remember they're more frequent than any other side antag, if you count the gateway and lavaland ruin too. Though, they'll just be sentient simple mob xenos after I got them nerfed hard. Don't buff them too hard.

TheSilverNuke commented:
>
>
>Though, they'll just be sentient simple mob xenos after I got them nerfed hard. Don't buff them too hard

I actually see this as a massive nerf towards xenos as a whole. If someone spots the queen when they just evolved it can result in an early game end for them.

Once I get more combat information from the eventual increased crew encounters with xenos I'll actually submit a second PR so it's not stun and done. This is just part one.

Coastfront commented:
"you cannot leave while fighting the arguably strongest team antagonist in the game" is what this is.

Mickyan commented:
> I actually see this as a massive nerf towards xenos as a whole. If someone spots the queen when they just evolved it can result in an early game end for them.

What this essentially does is give the crew a better fighting chance early on and forces the crew into a prolonged unwinnable curbstomp once that window of opportunity closes

In other words the GPS good, shuttle locking bad

IndieanaJones commented:
I've heard a lot of discussion about xenomorphs from many different people, and what I've come to understand, and believe myself, is that xeno's main issue is that were designed for an SS13 long gone in mind.  The stuns, the speed, all of it, it was made to be balanced back when people could outrun bullets.  Everything else changed, xenos never did.  If I were to change them right now, I'd do the following:
- Nerf the hunter's speed by a good margin, perhaps increase their health some to compensate
- Make neurotoxin work more like getting hit with a stun baton instead of instant down
- Decrease time on facehugger stun by a large margin, maybe expand upon the idea of some facewear tanking a hit and stunning the hugger, but it knocks the mask off
- Increase the strength of xeno structures to better reward base construction, but also lengthen the time it takes to create them so bases can't be sprung up immediately after losing one
- Allow xenomorphs to fire-carry
- Give them some objective and method to call the shuttle in the case they curb stomp the crew
- Bar xenos in the queenless state from entering vents or increase the penalty overall

Mind you, I'm no expert in the realm of xenos, but these are the thoughts I had on the matter.  If you really wanted to change them up, there were some ideas floating around about different humanoid species resulting in different xeno types, making them closer to source material and changing up how they play entirely, but it's not necessary at all.

TheSilverNuke committed:
Early Returns

TheSilverNuke committed:
Detonate Nuke / Remove News Report

Qustinnus commented:
sleeps still there

TheSilverNuke commented:
> 
> 
> sleeps still there

Yup. I asked on the coderbus if there was a viable alternative. I tried fiddling with it to get timers, but it didn't work.

Qustinnus commented:
Timers work fine, sleep() is not a valid option in this situation.

TheSilverNuke commented:
> 
> 
> Timers work fine, sleep() is not a valid option in this situation.

Can you message me on disc how you would write it instead?

TheSilverNuke committed:
Sleep is gone!!

TheSilverNuke committed:
Moved Procs and Check

No longer sitting on the queen directly

Qustinnus commented on code/modules/mob/living/carbon/alien/humanoid/queen.dm:
```diff
@@ -11,8 +11,27 @@
+			continue
+		living_humans++
+	if(living_humans < total_humans/10 && !nuking && !neutralized)
+		INVOKE_ASYNC(src, .proc/nuke_it_from_orbit)
```
does this need to be async

TheSilverNuke committed:
Missed one

Tlaltecuhtli commented on code/modules/antagonists/xeno/xeno.dm:
```diff
@@ -31,6 +31,54 @@
+
+/mob/living/carbon/alien/humanoid/royal/queen/Logout()
+	..()
+	addtimer(CALLBACK(src, .proc/logoutcheck), 12 SECONDS)//when logging out, we don't want to spam announcements with leaving and reconnecting.
```
should be 30 seconds at least because thats the timer not byond premium people have to wait to reconnect

TheSilverNuke committed:
Timer

I was going to add some more changes, and spent a few days fucking around with it but decided against it. I'm going to call it complete here if nobody objects.

AffectedArc07 commented on code/controllers/subsystem/ticker.dm:
```diff
@@ -156,10 +156,10 @@ SUBSYSTEM_DEF(ticker)
 			if(isnull(timeLeft))
 				timeLeft = max(0,start_at - world.time)
-			totalPlayers = LAZYLEN(GLOB.new_player_list)
+			totalPlayers = LAZYLEN(GLOB.new_player_list)
```
This file only has white space changes. Please undo.

TheSilverNuke replied:
That's strange, I didn't touch that file. 

![2020-05-14 05_53_01-ticker dm - tgstation - Visual Studio Code](https://user-images.githubusercontent.com/36066032/81931364-3638f180-95a7-11ea-8ee3-2f77271b12d9.png)

This is what I see.

AffectedArc07 commented on code/modules/antagonists/xeno/xeno.dm:
```diff
@@ -31,6 +31,56 @@
+	..()
+	addtimer(CALLBACK(src, .proc/logoutcheck), 40 SECONDS)//when logging out, we don't want to spam announcements with leaving and reconnecting.
+	SSshuttle.clearHostileEnvironment(src)
+	neutralized = TRUE
```
This will neutralise them if they logout, but not undo it if they log back in. Also I am pretty sure this should be in the xeno `Initialize` and `Destroy` procs.

TheSilverNuke replied:
I tested it and it will create a hostile environment when they're logged in, and when they log out it clears it just fine. I'm not sure what you mean when you say ``Initialize`` and ``Destroy`` however.

AffectedArc07 commented on code/modules/antagonists/xeno/xeno.dm:
```diff
@@ -31,6 +31,56 @@
+    nuke.safety = FALSE
+    nuke.explode()
+
+var/nuking = FALSE
```
I don’t think you meant to put this here, and if you did, we already have a global var for if the nuke is going off. Look at `nuclearbomb.dm` for some pointers.

TheSilverNuke replied:
Fixed

TheSilverNuke committed:
The last several updates are now null

Dumb

anconfuzedrock commented:
remember kids, if there's an emergency, that means you can't use the emergency shuttle.

PKPenguin321 commented:
\>spawn xenos to hurry a round to completion
\>actually it delays it by 30 minutes like a boss
please no

ExcessiveUseOfCobblestone commented:
The current xenos to my understanding are meant to END a round, not prolong it. Same as swarmers.

I'm fine with someone reworking them if they want to be seen as something like blob where you're meant to battle them, but I don't know if this is an appropriate change at the moment.

TheSilverNuke commented:
> 
> 
> The current xenos to my understanding are meant to END a round, not prolong it. Same as swarmers.
> 
> I'm fine with someone reworking them if they want to be seen as something like blob where you're meant to battle them, but I don't know if this is an appropriate change at the moment.

I fully intend on working on xenos, and as I mentioned in the description when asking about in on the servers it seems to be welcome. Additionally, this does in fact end the round, much better than if it weren't here. As it is if xenos kill the AI and nobody can get to a console the shuttle can't be called. This will not only allow for the shift to end in those circumstances but also allow other players who would normally never encounter a xeno to fight them. If the shuttle is kneejerked over even a hint of an infestation it would be difficult for anyone to locate them. This GPS signal also helps with that.

Rohesie commented:
I'll review the PR's code if the gameplay gets approved by some maintainer.

TheSilverNuke commented:
It literally isn't stale, I'm waiting for a maintainer to decide to merge it.

PKPenguin321 commented:
nobody wants to merge it i guess
which... is good

nemvar commented:
I know you put a lot of work into this but since there doesn't appear to be a maintainer willing to merge this, I will be closing this soon if it stays that way.
Sorry.

spookydonut commented:
Advanced merge begging
![image](https://user-images.githubusercontent.com/765153/83827906-f9799980-a711-11ea-9c4a-abef70025d1e.png)

TheSilverNuke commented:
Well that's one way to do things. I guess I can forget about rebalancing xenos.

spookydonut commented:
Maintainer shopping to get a pr merged isn't ok.

TheSilverNuke commented:
![2020-06-04 19_54_56-@Nemvar - Discord](https://user-images.githubusercontent.com/36066032/83828183-4cfdd000-a69d-11ea-8e6c-bf8df479fd8a.png)

In other words, "If you find someone willing to merge then by all means go ahead."

TheSilverNuke commented:
![2020-06-04 20_26_01-@spookydonut - Discord](https://user-images.githubusercontent.com/36066032/83830135-a405a400-a6a1-11ea-8253-0bcc343fe288.png)

What an amazing individual

wesoda25 commented:
based holy shit

AffectedArc07 commented:
You are aware of this point right?
![image](https://user-images.githubusercontent.com/25063394/83905211-31f09600-a759-11ea-841a-0002dcc18a2f.png)

A PR isnt guaranteed to be merged, and if its balance affecting it can often sit for ages while the implications are fully thought over. 
![image](https://user-images.githubusercontent.com/25063394/83905294-52b8eb80-a759-11ea-9972-ebbae83a73fe.png)
Also this kind of attitude isnt great either, not only the wording of your comment, but also instantly thumbs-downing any criticism towards you, as well as making sarc-y comments.
![image](https://user-images.githubusercontent.com/25063394/83905410-8a279800-a759-11ea-9fc0-838dc0374ba3.png)

> What an amazing individual

Grow up

Qustinnus commented:
lol DMing people to ask them to merge things is generally an epic gamer way to be donked

TheSilverNuke commented:
I'm sorry I wasn't aware that asking people to review my PR for consideration to merge was wrong. I am so sorry won't you ever forgive me?

AffectedArc07 commented:
![image](https://user-images.githubusercontent.com/25063394/83941888-d40a8f80-a7e6-11ea-93c0-c66638d0bc52.png)
Epic gamer move

Qustinnus commented:
imagine getting turned on by drawings of underage characters @TheSilverNuke

Canohaaaaaaa commented:
shots fired

Rohesie commented:
1) Don't bring stuff that might get us in trouble with the Github ToS.
2) This PR is closed and the event is past, let it rest.
